lib/alloc.c

```diff
@@ -71,3 +71,23 @@ xreallocarray(void *p, size_t nmemb, size_t size)
 	        log_get_progname(), strerror(errno));
 	exit(13);
 }
+
+char *
+xstrndup(const char *s, size_t max_length) {
+    size_t length = strlen(s);
```

Comment:
I think strndup(3) is not allowed to read more than `n` bytes in the source string.

Reply:
New approach is applied by direct usage of a new function: is_valid_user_name

Reply:
is_valid_user_name() is actually not new.  It was added in

```
commit 623d9e2ab324176c2c6e9fd731c00d46cc6c0ce1
Author: nekral-guest <nekral-guest@5a98b0ae-9ef6-0310-add3-de5d479b70d7>
Date:   Sun May 25 20:58:16 2008 +0000

            * libmisc/chkname.h, libmisc/chkname.c: check_group_name (resp.
            check_user_name) renamed to is_valid_user_name (resp.
            is_valid_group_name). is_valid_user_name and is_valid_group_name
            return a bool.
            * src/grpck.c, src/newusers.c, src/usermod.c, src/useradd.c,
            src/groupmod.c, src/pwck.c, src/groupadd.c: Use is_valid_user_name
            and is_valid_group_name, following above change.
            * libmisc/chkname.c: Avoid implicit conversion of chars to
            booleans. Add brackets and parenthesis.
```

However, you got rid of xstrndup(), so this can be closed.